### ISSUE TITLE
GetCurrentIpAddress calls cause severe slowdown in Fire Emblem: Three Houses

### ISSUE THREAD

**Issue by EmpressFiona** [Ryujinx/Ryujinx#2533]
### What's the issue you encountered?

Fire Emblem: Three Houses is a single-player game, but it has several asynchronous multiplayer features. Among them is the possibility, during weekly planning scenes (the 'calendar' screen), to look up what other players have done on the current week. This is entirely reliant on Nintendo's servers and completely useless to anyone playing on Ryujinx, but even when the game is set to Offline mode - indeed, you *can't* set it to Online, because you have to reach Nintendo's servers - the game will call `GetCurrentIpAddress` once per frame on the calendar screen (and *only* the calendar screen, at least as far as I tested), which slows the game to a complete crawl as it continually gets the IP it doesn't need and can't use, seemingly unable to remember it for more than a frame:

```
00:05:12.935 |I| HLE.OsThread.15 ServiceNifm GetCurrentIpAddress: Console's local IP is "192.168.1.199".
00:05:13.127 |I| HLE.OsThread.15 ServiceNifm GetCurrentIpAddress: Console's local IP is "192.168.1.199".
00:05:13.328 |I| HLE.OsThread.15 ServiceNifm GetCurrentIpAddress: Console's local IP is "192.168.1.199".
00:05:13.517 |I| HLE.OsThread.15 ServiceNifm GetCurrentIpAddress: Console's local IP is "192.168.1.199".
00:05:13.705 |I| HLE.OsThread.15 ServiceNifm GetCurrentIpAddress: Console's local IP is "192.168.1.199".
00:05:13.856 |I| HLE.OsThread.15 ServiceNifm GetCurrentIpAddress: Console's local IP is "192.168.1.199".
```

The rate at which the log fills with the message corresponds to the game's framerate. The game runs more or less perfectly (with some mild slowdown, perhaps due to shader compilation) in all other instances I've tested, and the calendar screen is hardly the most demanding, so there's no mistaking the culprit.

I asked on the Discord before coming here; they said it's a known issue, but I should report on GitHub anyway.

### How can the issue be reproduced?

Reproducing the issue should be as easy as getting to the calendar screen in Three Houses and looking in the console. (The `GetCurrentIpAddress` may be mixed in with other messages at first, but eventually it'll be the only thing in the console window, once per frame.)

In addition, it can be entirely fixed by commenting out segments of `IGeneralService.cs`, as seen below. I don't know what impact this would have on the LDN build, but since there's no capacity for the regular build to reach the network normally, I assume this is implemented because not doing so breaks some games - but since it can't reach it anyway, I don't see why it can't just say we're not connected. Well, I'll leave figuring out the exact solution to the devs; there could be other underlying issues here!

```c#
        public ResultCode GetCurrentIpAddress(ServiceCtx context)
        {
            /*(_, UnicastIPAddressInformation unicastAddress) = GetLocalInterface();

            if (unicastAddress == null)
            {*/
                return ResultCode.NoInternetConnection;
            /*}

            context.ResponseData.WriteStruct(new IpV4Address(unicastAddress.Address));

            Logger.Info?.Print(LogClass.ServiceNifm, $"Console's local IP is \"{unicastAddress.Address}\".");

            return ResultCode.Success;*/
        }
```

Apparently having it return `Success` immediately works just as well as `NoInternetConnection`, so maybe the culprit is the `GetLocalInterface` call or the `WriteStruct` call?

### Environment?

- Ryujinx version: 1.0.6982
- Game version: 1.2.0
- System Specs:
  - OS: Windows 10
  - CPU: i7-2600K (overclocked)
  - GPU: NVIDIA GTX 1660 SUPER
  - RAM: 16 GB
- Applied Mods : None. (I did apply a 60 FPS mod later, but it made no difference to the issue either way.)

### Additional context?

It's entirely possible this isn't more widely reported because it has a higher CPU impact on my machine than others. I'm using an old Sandy Bridge 2600K on an overclock, which is capable of running the game rather well in all other instances. I don't know if this is actually the case, though; maybe it's CPU-limited, or maybe the code is inefficient on Sandy Bridge?

The issue could be more of a bug with the *game* that ordinarily has no real impact on real hardware. The game keeps calling to figure out its own local IP address, despite the fact that it's not actually *supposed* to connect to the network at all; on real hardware it wouldn't matter, but here the game breaks. But then, it's also entirely possible it has to keep calling for the IP because something is broken in the pipeline somewhere?

On a real Switch, you'd be able to turn on Airplane mode to disable the Wi-Fi. Is it not possible to do that here, to just tell the Switch we're not connected to the network, so it shouldn't ask? Since networking isn't implemented, it shouldn't be any harm to at least provide the option for this call to return `NoInternetConnection`.

**Comment by jduncanator** [comment]
I have an inkling the issue behind this one is that we true when a game tries to determine if the internet connection is available: https://github.com/Ryujinx/Ryujinx/blob/00ce9eea620652b97b4d3e8cd9218c6fccff8b1c/Ryujinx.HLE/HOS/Services/Nifm/StaticService/IGeneralService.cs#L149

I'd be curious to see if changing this to return false would change the behaviour in this game. You can do that by changing [Line 22](https://github.com/Ryujinx/Ryujinx/blob/00ce9eea620652b97b4d3e8cd9218c6fccff8b1c/Ryujinx.HLE/HOS/Services/Nifm/StaticService/IGeneralService.cs#L22) to be false.

**Comment by jduncanator** [comment]
Tested that, doesn't seem to help. Will debug FE3H and see what the game is doing, and what it's expecting as a valid result.

**Comment by bylaws** [comment]
Is it the GetLocalInterface call itself that is taking the time? I imagine that has to do a fair bit of stuff to get the IP address, it might be the case that the call is quicker on HOS

**Comment by jduncanator** [comment]
> Is it the GetLocalInterface call itself that is taking the time? I imagine that has to do a fair bit of stuff to get the IP address, it might be the case that the call is quicker on HOS

It's likely where the performance loss comes from, but not the underlying cause. The game shouldn't be making a call to this IPC method thousands of times per frame. It's likely we return the wrong status somewhere, that triggers the game down this code path. I suspected it might be something to do with the IsAnyInternetRequestAccepted response we sent, but doesn't seem to be that.

I'll have to take a sift through the FE3H binary to see how and why this is being invoked so often. It's likely we just return the wrong error result code somewhere.

**Comment by JumpmanSr** [comment]
Hey, so I did some extensive looking into this and as far as I can tell, it's a glitch in the game itself and the only way to fix it would be a *hack* method kind of like what you described. I tried completely disabling any network emulation and it still makes the call regardless. 

The only solution I found working was to only get the local ip once and then cache and return that in subsequent calls. That increases FPS by quite a bit, but it introduces another issue. Since it's a hack it would have to be a toggle-able option. In doing so, the problem shifts from calling the GetLocalIP function to checking the config once a frame. While this improves fps quite a bit, it's still not ideal as it would only really target one game and and in-reality it would only be targeting a software issue, not an emulation issue. 

I might make a binary patcher if it's worth it, as far as I can tell, just nop'ing most of the function so it just returns no internet would be the most fps from my tests or maybe some kind of atmosphere base cheat code (but that requires making one for different build ids).

I'd love to hear if anyone has any input but I don't know if single-game hack options will be accepted so I haven't bothered to make a PR of the changes I've tried.

**Comment by jduncanator** [comment]
> Hey, so I did some extensive looking into this and as far as I can tell, it's a glitch in the game itself and the only way to fix it would be a _hack_ method kind of like what you described. I tried completely disabling any network emulation and it still makes the call regardless.
> 
> The only solution I found working was to only get the local ip once and then cache and return that in subsequent calls. That increases FPS by quite a bit, but it introduces another issue. Since it's a hack it would have to be a toggle-able option. In doing so, the problem shifts from calling the GetLocalIP function to checking the config once a frame. While this improves fps quite a bit, it's still not ideal as it would only really target one game and and in-reality it would only be targeting a software issue, not an emulation issue.
> 
> I might make a binary patcher if it's worth it, as far as I can tell, just nop'ing most of the function so it just returns no internet would be the most fps from my tests or maybe some kind of atmosphere base cheat code (but that requires making one for different build ids).
> 
> I'd love to hear if anyone has any input but I don't know if single-game hack options will be accepted so I haven't bothered to make a PR of the changes I've tried.

The most appropriate fix for this would probably be an exefs mod, which would patch the game binary to fix the issue. We could then link to that on the Game DB entry.

Ideally, this would be a bug in Ryujinx that we could fix, and I'm not quite ready to discount that yet, but if it's as you say and it's a bug in the game itself, then I don't really see any other option but to patch the binary.

**Comment by EmpressFiona** [comment]
I wouldn't mind trying out the exefs patch, at least! There's no point in having that call do anything anyway, because you can't connect to the servers in question.

I'm still unclear on why Ryujinx can't return that there isn't an Internet or network connection in non-LDN builds, since it's impossible for such builds to use networking features _anyway_. Since returning `NoInternetConnection` cuts off the bug, what's the point in returning an IP the game can't possibly use?

(I assume there is a technical reason, though, or at least that it would introduce more problems than it solves when it's a single game that is quite probably working just as it was coded.)

**Comment by jduncanator** [comment]
> I wouldn't mind trying out the exefs patch, at least! There's no point in having that call do anything anyway, because you can't connect to the servers in question.
> 
> I'm still unclear on why Ryujinx can't return that there isn't an Internet or network connection in non-LDN builds, since it's impossible for such builds to use networking features _anyway_. Since returning `NoInternetConnection` cuts off the bug, what's the point in returning an IP the game can't possibly use?
> 
> (I assume there is a technical reason, though, or at least that it would introduce more problems than it solves when it's a single game that is quite probably working just as it was coded.)

The "fix" suggested in the initial post isn't actually fixing the problem. All they've done is remove the (slow) call to `GetLocalInterface` and disabled the stub call. Behind the scenes, the game is still spamming this service call every frame, it's just now you don't know about it (because the stub log is gone), and it doesn't impact game speed, as the slow call to `GetLocalInterface` has been removed. The status code that gets returned isn't actually impacting the issue at all.

**Comment by JumpmanSr** [comment]
Hello everyone, firstly, sorry I didn't get back to this earlier I was a bit busy. Secondly, jd is right the first "fix" presented in the main post isn't actually a fix it's just returning nothing when the function asks for the IP. This would be bad as a "fix" as it would effectively be making the function do nothing; while it would help performance in FE3H, it could hurt emulation in other games.

That said, I actually figured out a "fix" to this. I'm going to submit a PR tomorrow that doesn't fundamentally change anything but prevents the slow down. 

Before: (Mainline / default settings / ran with vs debugger attached)
![before](https://user-images.githubusercontent.com/41526355/137677912-2edd6343-0287-445c-a9b5-f879fa2f5d94.png)
)

After: (Same build w/ my changes / default settings / ran with vs debugger attached)
![after](https://user-images.githubusercontent.com/41526355/137677982-e6660cd8-321b-4643-a675-3e3611371ee7.png)


Thank you for getting me interested in this, I hope my solution is good enough to be accepted but we'll see, I think it's about the best bet atm.
